projectq/backends/_ionq/_ionq_mapper.py

```diff
@@ -0,0 +1,74 @@
+from projectq.cengines import BasicMapperEngine
+from projectq.meta import LogicalQubitIDTag
+from projectq.ops import AllocateQubitGate, Command, DeallocateQubitGate, FlushGate
+from projectq.types import WeakQubitRef
+
+
+class BoundedQubitMapper(BasicMapperEngine):
+    def __init__(self, max_qubits):
+        super().__init__()
+        self._qubit_idx = 0
+        self.max_qubits = max_qubits
+
+    def _reset(self):
```

Comment:
Luckily I didn't need to worry about forcibly deallocating-- the important part was just resetting the index and validating it is within bounds. The plain dict takes care of the rest.